crates/precompile/src/bn128/substrate.rs

```diff
@@ -0,0 +1,144 @@
+        AffineG1::new(px, py)
+            .map(Into::into)
+            .map_err(|_| PrecompileError::Bn128AffineGFailedToCreate)
+    }
+}
+
+/// Creates a new `G2` point from the given Fq2 coordinates.
+///
+/// G2 points in BN128 are defined over a quadratic extension field Fq2.
+/// This function takes two Fq2 elements representing the x and y coordinates
+/// and creates a G2 point.
+///
+/// Note: The point at infinity which is represented as (0,0) is
+/// handled specifically because `AffineG2` is not capable of
+/// representing such a point.
+/// In particular, when we convert from `AffineG2` to `G2`, the point
+/// will be (0,0,1) instead of (0,1,0)
+#[inline]
+fn new_g2_point(x: Fq2, y: Fq2) -> Result<G2, PrecompileError> {
+    // TODO : Check whether or not we need these zero checks
+    let point = if x.is_zero() && y.is_zero() {
+        G2::zero()
+    } else {
+        G2::from(AffineG2::new(x, y).map_err(|_| PrecompileError::Bn128AffineGFailedToCreate)?)
+    };
+
+    Ok(point)
+}
+
+/// Reads a G1 point from the input slice.
+///
+/// Parses a G1 point from a byte slice by reading two consecutive field elements
+/// representing the x and y coordinates.
+///
+/// # Panics
+///
+/// Panics if the input is not at least 64 bytes long.
+#[inline]
+pub(super) fn read_g1_point(input: &[u8]) -> Result<G1, PrecompileError> {
+    let px = read_fq(&input[0..FQ_LEN])?;
+    let py = read_fq(&input[FQ_LEN..2 * FQ_LEN])?;
+    new_g1_point(px, py)
+}
+
+/// Encodes a G1 point into a byte array.
+///
+/// Converts a G1 point in Jacobian coordinates to affine coordinates and
+/// serializes the x and y coordinates as big-endian byte arrays.
+///
+/// Note: If the point is the point at infinity, this function returns
+/// all zeroes.
+#[inline]
+pub(super) fn encode_g1_point(point: G1) -> [u8; G1_LEN] {
+    let mut output = [0u8; G1_LEN];
+
+    if let Some(point_affine) = AffineG1::from_jacobian(point) {
+        point_affine
+            .x()
+            .to_big_endian(&mut output[..FQ_LEN])
+            .unwrap();
+        point_affine
+            .y()
+            .to_big_endian(&mut output[FQ_LEN..])
+            .unwrap();
+    }
+
+    output
+}
+
+/// Reads a G2 point from the input slice.
+///
+/// Parses a G2 point from a byte slice by reading four consecutive Fq field elements
+/// representing the two Fq2 coordinates (x and y) of the G2 point.
+///
+/// # Panics
+///
+/// Panics if the input is not at least 128 bytes long.
+#[inline]
+pub(super) fn read_g2_point(input: &[u8]) -> Result<G2, PrecompileError> {
+    let ba = read_fq2(&input[0..FQ2_LEN])?;
+    let bb = read_fq2(&input[FQ2_LEN..2 * FQ2_LEN])?;
+    new_g2_point(ba, bb)
+}
+
+/// pairing_check performs a pairing check on a list of G1 and G2 point pairs and
+/// returns true if the result is equal to the identity element.
+///
+/// Note: If the input is empty, this function returns true.
+/// This is different to EIP2537 which disallows the empty input.
```

Comment:
I guess this might be confusing to users since bn128 does something on empty inputs and bls12-381 does a different thing

Reply:
The comment you left is explanatory.